pinot-broker/src/main/java/org/apache/pinot/broker/routing/BrokerRoutingManager.java

```diff
@@ -214,87 +219,178 @@ private void processInstanceConfigChange() {
+    if ("true".equalsIgnoreCase(instanceConfigZNRecord.getSimpleField(Helix.QUERIES_DISABLED))) {
       return false;
     }
     return true;
   }
 
+  /**
+   * Excludes a server from the routing.
+   */
+  public synchronized void excludeServerFromRouting(String instanceId) {
+    LOGGER.warn("Excluding server: {} from routing", instanceId);
+    if (!_excludedServers.add(instanceId)) {
+      LOGGER.info("Server: {} is already excluded from routing, skipping updating the routing", instanceId);
+      return;
+    }
+    if (!_routableServers.contains(instanceId)) {
+      LOGGER.info("Server: {} is not enabled, skipping updating the routing", instanceId);
+      return;
+    }
+
+    // Update routing entry for all tables
+    long startTimeMs = System.currentTimeMillis();
+    Set<String> routableServers = new HashSet<>(_routableServers);
+    routableServers.remove(instanceId);
+    _routableServers = routableServers;
+    List<String> changedServers = Collections.singletonList(instanceId);
+    for (RoutingEntry routingEntry : _routingEntryMap.values()) {
+      try {
+        routingEntry.onInstancesChange(_routableServers, changedServers);
+      } catch (Exception e) {
+        LOGGER.error("Caught unexpected exception while updating routing entry when excluding server: {} for table: {}",
+            instanceId, routingEntry.getTableNameWithType(), e);
+      }
+    }
+    LOGGER.info("Excluded server: {} from routing in {}ms (updated {} routing entries)", instanceId,
+        System.currentTimeMillis() - startTimeMs, _routingEntryMap.size());
+  }
+
+  /**
+   * Includes a previous excluded server to the routing.
+   */
+  public synchronized void includeServerToRouting(String instanceId) {
+    LOGGER.info("Including server: {} to routing", instanceId);
+    if (!_excludedServers.remove(instanceId)) {
+      LOGGER.info("Server: {} is not previously excluded, skipping updating the routing", instanceId);
+    }
+    if (!_enabledServerInstanceMap.containsKey(instanceId)) {
+      LOGGER.info("Server: {} is not enabled, skipping updating the routing", instanceId);
+      return;
+    }
+
+    // Update routing entry for all tables
+    long startTimeMs = System.currentTimeMillis();
+    Set<String> routableServers = new HashSet<>(_routableServers);
+    routableServers.add(instanceId);
+    _routableServers = routableServers;
+    List<String> changedServers = Collections.singletonList(instanceId);
+    for (RoutingEntry routingEntry : _routingEntryMap.values()) {
+      try {
+        routingEntry.onInstancesChange(_routableServers, changedServers);
+      } catch (Exception e) {
+        LOGGER.error("Caught unexpected exception while updating routing entry when including server: {} for table: {}",
+            instanceId, routingEntry.getTableNameWithType(), e);
+      }
+    }
+    LOGGER.info("Included server: {} to routing in {}ms (updated {} routing entries)", instanceId,
```

Comment:
Why record durations with millitime here and nanotime elsewhere? They are implemented using  different clocks which makes the durations incomparable.

Reply:
My personal preference is to use nanotime to measure the duration, but there are some on-going discussion around which one should be favorable for performance perspective, and there is no conclusion yet. Since we already use millis in this class, I want to keep it consistent. Why would it cause incomparable durations? All the time we took here are in millis.

Reply:
There's no difference, they both call `clock_gettime`, `System.currentTimeMillis` with `CLOCK_REALTIME` and `System.nanoTime` with `CLOCK_MONOTONIC`, both clocks are in vDSO. I don't know what we need to do to settle this. Run a multithreaded microbenchmark on a range of servers?

The durations are incomparable because `System.currentTimeMillis` is subject to adjustment but `System.nanoTime` isn't. So if you record a number of milliseconds with `System.currentTimeMillis` it can be negative, smaller, the same, or larger than the same duration recorded with `System.nanoTime`.

Reply:
I see. We do use the same way to record the duration within the same scope (e.g. always use `millis` in the `BrokerRoutingManager`, and always use `nanos` in `BaseExponentialBackoffRetryFailureDetector` for the retry delay), and there is no comparison between them.
For the retry delay, I prefer using the nanos because it guarantees monotonically increasing. In the `BrokerRoutingManager`, since all the existing code already uses millis, I kept it the same

Reply:
The durations are tracked for the purpose of observability, and it's confusing that different clocks are used. Let's create a `Timer` SPI so users can choose and get consistency. No need to fix that here and now.